File: src/c/magma/api/stream/SingleHead.c
struct SingleHead<T> implements Head<T>{
	T value;
	boolean retrieved;
	public SingleHead(T value){
		this.value = value;
	}
	Option<T> next(){
		if(this.retrieved)return new None<>();
		this.retrieved = true;
		return new Some<>(this.value);
	}
}

File: src/java/magma/api/stream/SingleHead.java
package magma.api.stream;

import magma.api.option.None;
import magma.api.option.Option;
import magma.api.option.Some;

public final class SingleHead<T> implements Head<T> {
    private final T value;
    private boolean retrieved;

    public SingleHead(T value) {
        this.value = value;
    }

    @Override
    public Option<T> next() {
        if (this.retrieved) return new None<>();
        this.retrieved = true;
        return new Some<>(this.value);
    }
}
